### ISSUE TITLE
Proposal: Create a Readable stream based consumer

### ISSUE THREAD

**Issue by tizzo** [SOHU-Co/kafka-node#371]
Two important features of Kafka are message durability and strict message ordering (within a topic partition). As [others have noted](https://github.com/SOHU-Co/kafka-node/issues/338), this module's consumers convert kafka's fetch based API into a firehose event emitter that loads every message from the topic as fast as it can without regard for processing. That means that new messages are fired independently of whether the previous message is finished processing and this problem is compounded by the fact that [message decompression compromises the order the messages are delivered to event subscribers](https://github.com/SOHU-Co/kafka-node/issues/301).

It is currently possible to monitor the number of messages received and to [pause](https://github.com/SOHU-Co/kafka-node/blob/master/lib/consumer.js#L271) on the consumer to prevent a subsequent fetch, but this puts [quite a burden on calling code](https://github.com/SOHU-Co/kafka-node/issues/181#issuecomment-83931900) to constantly manage the `pause` and `resume` in the `message` event callback ([you can't fetch at all if the consumer is currently paused](https://github.com/SOHU-Co/kafka-node/blob/master/lib/consumer.js#L177)) and to do the bookkeeping to buffer messages and maintain their order (actually impossible if you're using compression).  This means that any system wanting to process messages in order, especially in circumstances where the destination is much slower than the source, needs to implement some kind of queuing/buffering and to do a lot of explicit pausing/un-pausing.

Node has a standard paradigm for processing I/O in sequence - The Almighty [Stream API](https://nodejs.org/api/stream.html).  The beauty of the stream API is that it is designed for strict ordering and buffering of fast sources and slow destinations. This makes it a _perfect_ fit for the Kafka protocol. An added bonus of this approach is that it makes this module work more consistently with the core node API and it provides a simple abstraction that handles all of the nasty details of a sort-order-enforcing queue and provides a clear answer on how to maintain order as messages are decompressed (pipe to a transform stream when appropriate). Best of all, consumers who want a simple emitter can still bind to `.on('data')` (and we could easily provide some magic to emit `.on('message')` as well so consumers don't need to update if they don't want to).

Another solvable issue with this module's current consumer firehose paradigm is that it makes it difficult to commit only messages that were successfully and completely processed. The fetching logic updates the internal offset list before emitting the fire-and-forget `message` event. This means that it's difficult to know whether you had a failure in your event processing and it means that you're likely to commit messages you didn't actually successfully read and process. One issue is that the `Consumer`'s [connect method binds the update offsets](https://github.com/SOHU-Co/kafka-node/blob/master/lib/client.js#L184) method to the [done event](https://github.com/SOHU-Co/kafka-node/blob/master/lib/client.js#L136) emitted by the `Client` upon successful fetch. That means that the commit functionality is essentially useless if you care about successfully processing events from a kafka queue in order and can't risk dropping messages. I'll admit you _can_ manage this yourself if you keep track of the queue mentioned above and use the `Client`'s [`sendOffsetCommitRequest`](https://github.com/SOHU-Co/kafka-node/blob/master/lib/client.js#L136) method directly, but again, this means doing a fair bit of book keeping yourself and it feels out of sync with the [guarantees](http://kafka.apache.org/documentation.html#intro_guarantees) and [use cases](http://kafka.apache.org/documentation.html#uses) that kafka is designed for. This is really annoying if you're using compressions as the you then also need to keep track of messages received out of order. The current implementation does a nice job for fire-and-forget pub-sub style use cases but much of Kafka differentiates itself based on the features beyond what you can do with redis pub-sub. This could also help to remove the need for the buggy [maxTickMessages](https://github.com/SOHU-Co/kafka-node/issues/303) option.

I propose that we implement a [Readable](https://nodejs.org/api/stream.html#stream_class_stream_readable) stream for the consumer classes. In the streams2+ paradigm, a readable stream emits the `readable` event when it is ready to start reading and the consumer calls read. At this point the readable stream should keep calling `this.push()` with new data until `this.push()` returns false - at that point the stream buffer is full and the stream should pause and wait for the `read()` method to be called again to perform the next fetch. If the Readable stream's [`highWaterMark`](https://nodejs.org/api/stream.html#stream_class_stream_readable) is coordinated with the kafka client's [`fetchMaxBytes`](https://nodejs.org/api/stream.html#stream_class_stream_readable) (which we can trivially do by default) this should allow us to easily ensure that messages are processed exactly in the correct order.

To address the problem with committing messages that have not been verifiably processed we could either include a callback in the message payload that will increment the offset for the topic in the consumer and/or implement a [writable stream](https://nodejs.org/api/stream.html#stream_class_stream_writable) that would receive messages and update the offset in the client. The auto-commit options and behaviors would remain the same but would not assume that every message emitted was properly processed (or at least provide a mode that does not assume that).

So the proposed syntax might be something like this:

``` javascript
var commitStream = client.createCommitStream({
  // This would commit every 100 *successfully processed* messages.
  autoCommitMsgCount: 100,
  // This would commit the most recent *successfully processed* offset every second.
  autoCommitIntervalMs: 1000,
});
consumer.stream
  .pipe(through2.obj(function(message, enc, cb) {
    console.log(message.data);
    // Do some stuff with the data
    this.push(message);
    cb();
  })))
  .pipe(commitStream);

```

Alternatively, if we gave the message a commit callback it could look like this:

``` javascript
consumer.stream
  .pipe(through2.obj(function(message, enc, cb) {
    console.log(message.data);
    // Do some stuff with the data
    this.push(message);
    // We successfully processed this message.
    message.commit();
    cb();
  })))
```

I think the former commit model is probably a better one as the writable stream can easily maintain state and ensure that messages aren't skipped, emitting an error if one is.

**Comment by tb01923** [comment]
+1 I am using this project as an under pinning of a lightweight (emphasis _light_) streaming framework (e.g. samza/storm/spark) for NodeJs, and would like to be able to work this way.  My focus is really creating an expressive systems for the application engineers to utilize without them having to worry about the complexities underpinning that solution.

**Comment by sehod** [comment]
💯 agree that a consumer implementation based on Readable is needed.

**Comment by estliberitas** [comment]
cc @hyperlink - I'm not ready to work on this, but it sounds very interesting

**Comment by argh** [comment]
Couldn't agree more.  I ran into the exact issues described by @tizzo and suffered the pain of implementing a secondary in-memory queue as well.  The manual pause/resume is actually not only inefficient and an implementation burden, but in high traffic conditions you can't rely on your own counters for backpressure because the internal consumer might have already buffered a huge number of messages.  A stream API is clearly the right approach.

**Comment by Foo42** [comment]
I spiked out a stream based reader and began a commit stream about 6 months back for a project at work. It seemed to work under basic testing. Ss you suggested we use the return from push to pause the consumer etc. I've just returned to the code as we have an interest in productionising it. If it all works out (with the commit stream etc) I'll feedback, maybe see if I can get agreement to publish our implementation.

**Comment by pcothenet** [comment]
Where do we stand on that one? Has anyone started giving it a shot?

**Comment by Foo42** [comment]
I've got some code I've spiked out with both a readable kafka topic stream and a writeable kafka commit stream. It seems to work, with back pressure / demand pausing and resuming the kafka consumer. I'll try and get it cleaned up a bit and open a PR. It can be a starting point for discussion anyway

**Comment by gbarton** [comment]
Any updates on this one? Having the commit on the message and guarantee of order per partition is exactly how I am trying to use kafka.

**Comment by tizzo** [comment]
I'm currently working on a PR - hoping to have some code to show in the
next day or two.

**Comment by jwthomp** [comment]
Has this gone anywhere?

**Comment by itamarwe** [comment]
+1

**Comment by barrotsteindev** [comment]
I'll be more than happy to try and give it a shot.

**Comment by tinder-dyakobian** [comment]
This is how I do it https://gist.github.com/tinder-dyakobian/6d06095acac7f7e22255b526cc2d3dd3

**Comment by lliss** [comment]
This is implemented in https://github.com/SOHU-Co/kafka-node/pull/695

**Comment by hyperlink** [comment]
Published to npm as v2.1.0

**Comment by tizzo** [comment]
*YEEEEESSSS*

Thanks everyone for pushing this over the finish line and making my dream a reality.

![van damme dance](https://media1.giphy.com/media/FBzqZGthkW6KQ/giphy.gif)

![motivational dance](https://media0.giphy.com/media/NeKB8ZR67TuV2/giphy.gif)
